File: bindings/java/src/ndwk/Sample.java
package ndwk;

import java.io.File;
import java.io.FileInputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class Sample {
    public static void main(String[] args) throws Exception {
        System.out.println("=== ndwk Java (Panama FFM) Demo ===");

        String wavPath = args.length > 0 ? args[0] : "test/ja/ja_033.wav";
        String modelsDir = args.length > 1 ? args[1] : "models";

        if (!new File(wavPath).exists()) {
            System.err.println("WAV file not found: " + wavPath);
            return;
        }

        // 1. WAV 読み込み (16kHz 16-bit Mono -> float[])
        float[] samples = loadWavPcm16(wavPath);
        System.out.printf("Loaded: %s (%.2fs, %d samples)%n", wavPath, samples.length / 16000.0, samples.length);

        // 2. エンジン生成 (try-with-resources で自動解放)
        try (var engine = new NdwkEngine(
                modelsDir,
                NdwkLang.JA,
                true,
                partial -> {
                    System.out.printf("\r\033[32m[Partial(Java)]\033[0m: %s\033[K", partial);
                    System.out.flush();
                },
                (lang, text) -> {
                    System.out.printf("\r\033[33m[Final(Java / %s)]\033[0m: %s\033[K%n", lang, text);
                    System.out.flush();
                }
        )) {
            System.out.println("Streaming audio to ndwk...\n");

            // 3. 疑似ストリーミング (100ms = 1600サンプルずつ投入)
            int chunkSize = 1600;
            for (int i = 0; i < samples.length; i += chunkSize) {
                int len = Math.min(chunkSize, samples.length - i);
                engine.feedAudio(samples, i, len);
                Thread.sleep(50);
            }

            // 4. フラッシュ
            engine.flush();
            Thread.sleep(3000); // 最終結果が upcall で返ってくるまで少し待つ
            System.out.println("\nDone.");
        }
    }

    private static float[] loadWavPcm16(String path) throws Exception {
        try (var fis = new FileInputStream(path)) {
            byte[] header = fis.readNBytes(12);
            if (!"RIFF".equals(new String(header, 0, 4)) || !"WAVE".equals(new String(header, 8, 4))) {
                throw new IllegalArgumentException("Not a valid WAV file");
            }

            byte[] chunkHeader = new byte[8];
            while (fis.read(chunkHeader) == 8) {
                String chunkId = new String(chunkHeader, 0, 4);
                int chunkSize = ByteBuffer.wrap(chunkHeader, 4, 4).order(ByteOrder.LITTLE_ENDIAN).getInt();

                if ("data".equals(chunkId)) {
                    byte[] raw = fis.readNBytes(chunkSize);
                    int sampleCount = chunkSize / 2;
                    float[] samples = new float[sampleCount];
                    ByteBuffer bb = ByteBuffer.wrap(raw).order(ByteOrder.LITTLE_ENDIAN);
                    for (int i = 0; i < sampleCount; i++) {
                        samples[i] = bb.getShort() / 32768.0f;
                    }
                    return samples;
                } else {
                    fis.skipNBytes(chunkSize);
                }
            }
            throw new IllegalArgumentException("Data chunk not found in WAV");
        }
    }
}

File: bindings/python/sample.py
#!/usr/bin/env python3
"""ndwk Python binding sample — WAV file transcription."""

import sys
import wave
import array
from ndwk import NdwkEngine, NdwkLang

def main() -> None:
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <wav_file> <models_dir>")
        sys.exit(1)

    wav_path = sys.argv[1]
    models_dir = sys.argv[2]

    print("=== ndwk Python (ctypes) Demo ===")

    # WAV読み込み (16kHz mono s16le)
    with wave.open(wav_path, "rb") as wf:
        assert wf.getnchannels() == 1, "Mono WAV required"
        assert wf.getsampwidth() == 2, "16-bit WAV required"
        assert wf.getframerate() == 16000, "16kHz WAV required"
        n_frames = wf.getnframes()
        raw = wf.readframes(n_frames)

    # s16le → float32 [-1.0, 1.0]
    shorts = array.array("h", raw)
    samples = array.array("f", (s / 32768.0 for s in shorts))
    duration = len(samples) / 16000.0
    print(f"Loaded: {wav_path} ({duration:.2f}s, {len(samples)} samples)")

    # コールバック
    def on_partial(text: str) -> None:
        print(f"\r\x1b[32m[Partial]\x1b[0m: {text}\x1b[K", end="")

    def on_final(lang : NdwkLang, text: str) -> None:
        print(f"\r\x1b[32m[Final({lang.name})]\x1b[0m: {text}\x1b[K")

    # エンジン起動
    engine = NdwkEngine(
        models_dir=models_dir,
        lang=NdwkLang.JA,
        on_partial=on_partial,
        on_final=on_final,
    )

    # ストリーミング送信 (1600サンプル = 100ms ずつ)
    print("Streaming audio to ndwk...")
    chunk_size = 1600
    for i in range(0, len(samples), chunk_size):
        engine.feed_audio(samples[i : i + chunk_size])

    # 残りをflush
    engine.flush()
    engine.close()

    print("\nDone.")


if __name__ == "__main__":
    main()
